src/lib.rs

```diff
@@ -0,0 +1,534 @@
+    /// This function should be called in every iteration of a memtest
+    ///
+    /// To minimize overhead, the time is only checked at specific checkpoints.
+    /// The algorithm makes a prediction of how many iterations will be done during `checking_interval`
+    /// and sets the checkpoint based on the prediction.
+    ///
+    /// The algorithm also determines whether it is likely that the test will be completed
+    /// based on `work_progress` and `time_progress`.
+    /// If it is likely that the test will be completed, `checking_interval_ns` is scaled up to be more
+    /// lenient and reduce overhead.
+    fn check(&mut self) -> Result<(), MemtestError> {
+        let Some(test_instance) = &mut self.curr_test else {
+            panic!("Attempted to check timeout without initialization");
+        };
+        if test_instance.completed_iter < test_instance.checkpoint {
+            test_instance.completed_iter += 1;
+            return Ok(());
+        }
+
+        let curr_time = Instant::now();
+        if curr_time >= self.deadline {
+            return Err(MemtestError::Timeout);
+        }
+
+        let work_progress =
+            test_instance.completed_iter as f64 / test_instance.expected_iter as f64;
+        // Would be nice to log every percent, but checking it every iteration causes huge overhead
+        // TODO: This current method of logging progress is quite limited, especially for multithreading
+        if test_instance.completed_iter % (test_instance.expected_iter / 100) == 0 {
+            trace!("Progress: {:.0}%", work_progress * 100.0);
+        }
+
+        let test_elapsed = curr_time - test_instance.test_start_time;
+        let time_progress =
+            test_elapsed.div_duration_f64(self.deadline - test_instance.test_start_time);
+        // TODO: Consider having a max for `checking_interval` to have a reasonable timeout guarantee
+        if work_progress > time_progress {
+            test_instance.checking_interval = test_instance.checking_interval.saturating_mul(2);
+        }
+
+        let avg_iter_duration = test_elapsed.div_f64(test_instance.completed_iter as f64);
+        let iter_per_interval = test_instance
+            .checking_interval
+            .div_duration_f64(avg_iter_duration) as u64;
+        test_instance.checkpoint = test_instance.completed_iter + iter_per_interval;
+
+        test_instance.num_checks_completed += 1;
+        test_instance.completed_iter += 1;
+        Ok(())
+    }
+}
+
+// TODO: Rethink options for handling mlock failure
+//       The linux memtester always tries to mlock,
+//       If mlock returns with ENOMEM or EAGAIN, it resizes memory.
+//       If mlock returns with EPERM or unknown error, it moves forward to tests with unlocked memory.
+//       It is unclear whether testing unlocked memory is something useful
+// TODO: Check for timeout, decrementing memory size can take non trivial time
+fn memory_resize_and_lock(
```

Comment:
Rather than this and the next function, I think it would be less ceremony to just add this to the top of the file:

```Rust
#[cfg(windows)]
use windows::{memory_resize_and_lock, memory_unlock};
#[cfg(unix)]
use unix::{memory_resize_and_lock, memory_unlock};
```